test: raft: fix the topology failure recovery test flakiness

emdroid opened the pull request:
Setting the error condition for all nodes in the cluster to avoid having to check which one is the coordinator. This should make the test more stable and avoid the flakiness observed when the coordinator node is the one that got the error condition injected.

Randomizing the retrieved running servers to reproduce the issue more frequently and to avoid making any assumptions about the order of the servers.

Note that only the "raft_topology_barrier_fail" needs to run on a non-coordinator node, the other error "stream_ranges_fail" can be injected on any node (including the coordinator).

Fixes: #18614

emdroid commented on test/topology/test_topology_failure_recovery.py:
```diff
@@ -4,73 +4,105 @@
+        random (especially when initialized in parallel).
+    """
+    servers = await manager.running_servers()
+    random.shuffle(servers)
```
The randomization has been done to make sure the problem is reproducible with reasonable rate (e.g. when run with "--repeat 100"). The actual fix is the `choose_non_coordinator` so the randomization could be removed eventually, but I think that it would actually make sense to keep it so that there can't be any assumptions made about the retrieved order of servers.

I'm also thinking whether it would make sense to even randomize the `manager.running_servers()` response - it would have the potential to break many tests, but in particular that would be the ones that might have a similar problem like this one (making assumptions about the server ordering).

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10298/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10298/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology/test_topology_failure_recovery
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10298/)
#### Build Details:
- Duration: 1 hr 1 min
- Builder: spider8.cloudius-systems.com

kbr-scylla commented on test/topology/test_topology_failure_recovery.py:
```diff
@@ -4,73 +4,105 @@
 # SPDX-License-Identifier: AGPL-3.0-or-later
```
Re commit message

> Closes https://github.com/scylladb/scylladb/issues/18614

Should be `Fixes`
in particular:
`Fixes scylladb/scylladb#18614`

kbr-scylla commented on test/topology/test_topology_failure_recovery.py:
```diff
@@ -4,73 +4,105 @@
+async def choose_non_coordinator(manager: ManagerClient, servers):
+    """ Return a server that is not the coordinator.
+    """
+    coord_id = await get_topology_coordinator(manager)
```
BTW I looked into implementation of this and I see some problems with it (cc @tgrabiec)
- it actually returns Raft leader for the group, not the topology coordinator; the topology coordinator does not need to be the leader -- it follows soon after, but generally it does not start "atomically" with the leader
- it takes a `group_id` parameter so the implementation  appears to be prepared for the future -- when there will be multiple groups -- but it routes the request to shard 0. But not all groups will live on shard 0

But these are minor problems and fixing them is out of scope of this PR.

Still I kind of dislike the approach of "finding coordinator" in general, because the coordinator (or the raft leader) can in theory migrate at any time. The logic of "find coordinator -> do something to the coordinator (or not-coordinator)" assumes that the coordinator won't change before the "do something", which is in general wrong. 

In the case of our test, a more robust approach -- and I think it would result in actually much less code -- would be injecting the failure on all servers. Then no matter who the coordinator is, it will work.

But I don't reject this approach, you did the randomization to reproduce the problem often and verified it no longer reproduces with this approach, so if it works I can accept it.

emdroid replied:
Yes, it looks like at least for the test the leader (/coordinator) don't migrate - even in the failing test in #18614 the `s[1]` was the leader since the very beginning, just that it was the second server in the list and not the first one.

Without the randomization I wasn't even able to reproduce the issue despite using `--repeat 100`, but with randomization it is reproduced easily (and after the fix the `--repeat 100` is all green again).

Injecting the failure on all servers would work too I suppose.

kbr-scylla reviewed (commented):
Minor fix needed in commit message, otherwise LGTM

kbr-scylla commented on test/topology/test_topology_failure_recovery.py:
```diff
@@ -4,73 +4,100 @@
+    servers = await get_running_servers(manager)
     s = await manager.server_add(start=False)
-    await manager.api.enable_injection(servers[1].ip_addr, 'raft_topology_barrier_fail', one_shot=True)
+    await inject_error_on(manager, "raft_topology_barrier_fail", servers)
```
Hm, this is tricky. I see a potential problem here.

The error is injected on all servers -- good. But then we know that one server will not stumble on it -- the coordinator. Which means that it will remain armed on the coordinator. Now suppose the coordinator changes. The previous coordinator has still the injection armed. This might potentially interfere with the remainder of this test -- e.g. one of these operations might stumble on this injection, not on the other ones, and either fail with unexpected error, or pass but not because it tested what it was supposed to test.

To eliminate this possibility and be 100% safe, I think you should explicitly disable this injection after this case finishes (i.e. before `#rebuild failure` below)

emdroid replied:
Yes, that sounds right. Added the explicit removal of the error injection.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10332/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10332/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology/test_topology_failure_recovery
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10332/)
#### Build Details:
- Duration: 57 min
- Builder: spider7.cloudius-systems.com

emdroid commented:
New version:
* rebased
* added the explicit removal of the error injection for the `raft_topology_barrier_fail`, as it will not trigger at the coordinator and would remain armed otherwise

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10346/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10346/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology/test_topology_failure_recovery
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10346/)
#### Build Details:
- Duration: 2 hr 37 min
- Builder: i-0c3d56e05e2029dcf (m5ad.8xlarge)
